File: src/part2/recode/java/Recoder.java
package part2.recode.java;

import java.io.*;

@SuppressWarnings("WeakerAccess")
public class Recoder {

    private final String charsetInput;

    private final String charsetOutput;

    public Recoder(String charsetInput, String charsetOutput) {
        this.charsetInput = charsetInput;
        this.charsetOutput = charsetOutput;
    }

    public int recode(InputStream in, OutputStream out) throws IOException {
        try (InputStreamReader reader = new InputStreamReader(in, charsetInput)) {
            try (OutputStreamWriter writer = new OutputStreamWriter(out, charsetOutput)) {
                int sym = reader.read();
                int count = 0;
                while (sym != -1) {
                    writer.write(sym);
                    count++;
                    sym = reader.read();
                }
                return count;
            }
        }
    }

    public int recode(String inputName, String outputName) throws IOException {
        try (FileInputStream inputStream = new FileInputStream(inputName)) {
            try (FileOutputStream outputStream = new FileOutputStream(outputName)) {
                return recode(inputStream, outputStream);
            }
        }
    }
}

File: src/part2/recode/kotlin/Recoder.kt
package part2.recode.kotlin

import java.io.*

class Recoder(private val charsetInput: String, private val charsetOutput: String) {

    @Throws(IOException::class)
    fun recode(`in`: InputStream, out: OutputStream): Int {
        InputStreamReader(`in`, charsetInput).use { reader ->
            OutputStreamWriter(out, charsetOutput).use { writer ->
                var sym = reader.read()
                var count = 0
                while (sym != -1) {
                    writer.write(sym)
                    count++
                    sym = reader.read()
                }
                return count
            }
        }
    }

    @Throws(IOException::class)
    fun recode(inputName: String, outputName: String): Int {
        FileInputStream(inputName).use { inputStream ->
            FileOutputStream(outputName).use { outputStream ->
                return recode(inputStream, outputStream)
            }
        }
    }
}
